File: 2020/c/5.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "aoc.h"

static int parse(char *str)
{
    int i, d;

    for (i = 0, d = 0; i < 10; i++) {
        if (str[i] == 'B' || str[i] == 'R')
            d += (1 << (9 - i));
    }
    return d;
}

int main(void)
{
    int id, maxid, myid;
    char line[12];

    hinit();
    maxid = myid = 0;

    while (fgets(line, 12, stdin) != NULL) {
        id = parse(line);
        hintinsert(id, id);

        /* Part A */
        if (id > maxid)
            maxid = id;
    }

    /* Part B */
    for (id = 1; id < 1023; id++) {
        if (hintlookup(id - 1) != NULL
            && hintlookup(id + 1) != NULL
            && hintlookup(id) == NULL)
            myid = id;
    }
    hfree(free);

    printf("A) the highest seat id is %d\n", maxid);
    printf("B) the missing seat id is %d\n", myid);

    return 0;
}

File: 2020/zig/src/5.zig
const std = @import("std");

fn parse(line: []u8) u10 {
    var d: u10 = 0;
    var i: u4 = 10;
    for (line) |c| {
        i -= 1;
        if (c == 'B' or c == 'R') {
            d += (@as(u10, 1) << i);
        }
    }
    return d;
}

pub fn main() anyerror!void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = &arena.allocator;
    
    var map = std.AutoHashMap(u10, u10).init(allocator);
    var id: u10 = undefined;
    var na: u10 = 0;

    var stdin = std.io.getStdIn().reader();
    var in = std.io.bufferedReader(stdin).reader();
    var buf: [256]u8 = undefined;
    while (try in.readUntilDelimiterOrEof(&buf, '\n')) |line| {
        id = parse(line);
        try map.put(id, id);

        // Part A
        if (id > na) {
            na = id;
        }
    }

    // Part B
    var nb: u10 = undefined;
    id = 1;
    while (id < 1023) : (id += 1) {
        if (map.get(id - 1) != null and
            map.get(id + 1) != null and
            map.get(id) == null) {
            nb = id;
        }
    }

    std.debug.print("A) the highest seat id is {d}\n", .{na});
    std.debug.print("B) the missing seat id is {d}\n", .{nb});
}
